### PR TITLE
feat: add text filter on org member list

### LINKED ISSUE #17644: [CAL-4676] Improved filters
An advanced filter as such rather than just a X === Y

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-4676](https://linear.app/calcom/issue/CAL-4676/improved-filters)</sub>

### PR DESCRIPTION
**Pull request by eunjae-lee**
## What does this PR do?

- Fixes #17644 (GitHub issue number)
- Fixes CAL-4676 (Linear issue number - should be visible at the bottom of the GitHub issue description)

### 1. New Text filter

This PR add a text filter for text attributes on organization member list. These are the text operators: `Is`, `Is not`, `Contains`, `Does not contain`, `Starts with`, `Ends with`, `Is empty`, `Not empty`

(Just like before) it works on both client and server side:
- `filterFn` runs `textFilter` function to filter the rows that are on browser side.
- And then the search states (`?activeFilters=`) re-triggers `trpc.viewer.organizations.listMembers.useInfiniteQuery()` call, and its handler uses the text filters in the where clause correctly.

### 2. Bug fix on initial load

This PR also fixes an existing bug that the initial load with the `?activeFilters=...` params didn't work and the user had to re-apply filters on the browser side.

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] N/A
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

https://github.com/user-attachments/assets/e3999ca8-e48b-433c-9724-6eb3c8d03039

- The text filter works only on text attributes.
- With this change, we are now able to introduce a numeric filter as well (in a separate PR).

### CODE REVIEW

File: packages/features/users/components/UserTable/UserListTable.tsx [hunk 1]
```diff
@@ -117,18 +119,30 @@ export function UserListTable() {
...
+      }))
+      .filter((filter) => {
+        // Keep the empty array state on the client side,
+        // but do not send it to the actual query.
+        if (Array.isArray(filter.value) && filter.value.length === 0) {
+          return false;
+        }
+        return true;
+      });
+  }, [filtersSearchState.activeFilters]);
```
Comment (eunjae-lee) [inline packages/features/users/components/UserTable/UserListTable.tsx:137]: We already have `?activeFilters=` on the search state (URL), but kept `columnFilters` separately from `onColumnFilterChange` below. I think it's unnecessary to keep two sources, so I've made `columnFilters` out of `useFiltersSearchState()`.

File: packages/ui/components/data-table/DataTableFilter.tsx
Comment (eunjae-lee) [inline packages/ui/components/data-table/DataTableFilter.tsx:1]: It seems this file is not being in use.

File: packages/ui/components/data-table/filters/utils.ts [hunk 2]
```diff
@@ -15,3 +18,111 @@ export const filtersSearchParams = {
 export function useFiltersSearchState() {
   return useQueryStates(filtersSearchParams);
 }
+
+export type FiltersSearchState = ReturnType<typeof useFiltersSearchState>[0];
+export type SetFiltersSearchState = ReturnType<typeof useFiltersSearchState>[1];
+export type ActiveFilter = NonNullable<FiltersSearchState["activeFilters"]>[number];
+
+export const textFilter = (cellValue: string, filterValue: TextFilterValue) => {
```
Comment (eunjae-lee) [inline packages/ui/components/data-table/filters/utils.ts:26]: This function is used with tanstack table to filter the existing rows on the client side.

File: packages/ui/components/data-table/filters/utils.ts [hunk 3]
```diff
@@ -15,3 +18,111 @@ export const filtersSearchParams = {
...
+    "type" in filterValue &&
+    filterValue.type === "text"
+  );
+};
+
+export const isSelectFilterValue = (filterValue: unknown): filterValue is SelectFilterValue => {
+  return Array.isArray(filterValue) && filterValue.every((item) => typeof item === "string");
+};
+
+export function makeWhereClause(columnName: string, filterValue: FilterValue) {
```
Comment (eunjae-lee) [inline packages/ui/components/data-table/filters/utils.ts:62]: This is a utility function for prisma for the backend to build proper where clause.

File: packages/features/users/components/UserTable/UserListTable.tsx [hunk 4]
```diff
@@ -117,18 +119,15 @@ export function UserListTable() {
   const [debouncedSearchTerm, setDebouncedSearchTerm] = useState("");
   const [isDownloading, setIsDownloading] = useState(false);
 
-  const [columnFilters, setColumnFilters] = useState<ColumnFiltersState>([]);
+  const columnFilters = useColumnFilters();
```
Comment (eunjae-lee) [inline packages/features/users/components/UserTable/UserListTable.tsx:122]: We already have `?activeFilters=` on the search state (URL), but kept `columnFilters` separately with `onColumnFilterChange` below. I think it's unnecessary to keep two sources, so I've made `columnFilters` out of `useFiltersSearchState()`.

File: packages/ui/components/data-table/filters/MultiSelectFilterOptions.tsx [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+export function MultiSelectFilterOptions({
+  column,
+  filterValue,
+  setFilterValue,
+  removeFilter,
+}: MultiSelectFilterOptionsProps) {
+  const { t } = useLocale();
+
+  return (
```
Comment (eunjae-lee) [inline packages/ui/components/data-table/filters/MultiSelectFilterOptions.tsx:32]: This part is extracted from `packages/ui/components/data-table/filters/index.tsx`. The markups are almost the same, but only `onSelect` has changed a bit.

**Comment by zomars** [comment]
Hey @eunjae-lee from the demo is kinda hard to see but it seems like some filters are being applied without hitting the "apply" button. Are you hitting the Enter key? If not can we enforce consistency so we only apply them when hitting the button or the Enter key?

**Comment by eunjae-lee** [comment]
> Hey @eunjae-lee from the demo is kinda hard to see but it seems like some filters are being applied without hitting the "apply" button. Are you hitting the Enter key? If not can we enforce consistency so we only apply them when hitting the button or the Enter key?

Oh the video wasn't clear. Yes, you're right. It's either enter key or hitting the button, because it's a button with type="submit"

**Comment by keithwillcode** [comment]
@sean-brydon Can you take a look at this one, please?

File: packages/trpc/server/routers/viewer/organizations/listMembers.handler.ts [hunk 6]
```diff
@@ -81,9 +82,7 @@ export const listMembersHandler = async ({ ctx, input }: GetOptions) => {
               attribute: {
                 id: filter.id,
               },
-              value: {
-                in: filter.value,
-              },
+              ...makeWhereClause("value", filter.value),
```
Comment (Udit-takkar) [inline packages/trpc/server/routers/viewer/organizations/listMembers.handler.ts:85]: Filtering by text shouldn't be case sensitive

https://github.com/user-attachments/assets/9f833e55-d394-4e4f-8c90-385bef236923

File: packages/ui/components/data-table/filters/TextFilterOptions.tsx [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+              disabled={form.formState.isSubmitting}
+              onClick={() => removeFilter(column.id)}>
+              Clear
+            </Button>
+            <Button
+              type="submit"
+              color="primary"
+              loading={form.formState.isSubmitting}
+              disabled={form.formState.isSubmitting}>
+              Apply
```
Comment (Udit-takkar) [inline packages/ui/components/data-table/filters/TextFilterOptions.tsx:103]: NIT: use i18n

**Review by sean-brydon — DISMISSED** [review]
Tested all works well: 

However, i do believe @Udit-takkar's comment is blocking. We should not care about the case of the text on the text inputs. 

I also wonder if we should address the NIT to move the makeWhereQuery to elsewhere

**Comment by eunjae-lee** [comment]
Moving the whole data table related files to /features. Some stuff is broken now but I will fix them after lunch.

**Review by Udit-takkar — COMMENTED** [review]
LGTM. 
There are some merge conflicts

**Comment by eunjae-lee** [comment]
> LGTM. There are some merge conflicts

@Udit-takkar fixed :)

**Review by Udit-takkar — APPROVED** [review]

**Pull request merged**